Add course planning to for group admins to tentatively schedule instructors

jxjj opened the pull request:
## Summary

This allow users with `edit planned courses` permission to tentatively plan instructors and TA's for future terms.

Closes #70 

## Screenshots
![ScreenShot 2023-12-14 at 11 30 40@2x](https://github.com/UMN-LATIS/bluesheet/assets/980170/38fb6202-0dfd-4a49-b48b-6a1f9858767c)
![ScreenShot 2023-12-14 at 11 31 31@2x](https://github.com/UMN-LATIS/bluesheet/assets/980170/a01d07a3-5ef6-475a-b5a1-9416eacff66c)
![ScreenShot 2023-12-14 at 11 32 42@2x](https://github.com/UMN-LATIS/bluesheet/assets/980170/22fc248c-e27a-4758-96b0-27900e876c3c)

## Details

Terminology in this PR:
-  **course** (e.g. HIST-1001). 
-  **section** is a course in a term with a section number. It has a unique `CLASS_NUMBER`. A **course** contains many **sections**. 
- **enrollment** is a user in a section with role (Primary Instructor, Teaching Assistant) . A **section** may contain many enrollments.
- A **section** is **published** if it's in the SIS (bandaid). If it's only stored locally, it's considered **unpublished**.

Notable Changes:
- adds `course_sections` and `enrollments` tables and models to our database for tracking tentative unpublished class section data. 
- adds controllers and routes for CRUD operations on course_sections and enrollments for a given group. A single route will return info about about published and unpublished sections in the same shape.
- splits pinia course history store into several normalized stores
- adds vuedraggable package to handle drag and drop
- refactors CoursePlanningPage and adds components for unpublished sections

cmcfadden commented on app/Enrollment.php:
```diff
@@ -0,0 +1,24 @@
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class Enrollment extends Model {
```
How much would it suck to change this to something other than `enrollments`? Just wondering about confusion down the line if we do more enrollment-centric features (section count / enrollment size / etc)

jxjj replied:
It seems doable. Do you want to do it now, or just wondering for a future change? Have something in mind? 

Canvas calls it "enrollment," regardless of role: student, instructor, ta. Blackboard calls it "course membership," but we already have members for groups so that might be confusing. Is there a standard UMN or PeopleSoft term that's used?

cmcfadden replied:
As far as I know, they're all called instructors in peoplesoft, and then they have different roles. I can't think of another term in use at the U, and glancing through the LDW I don't see anything. It's probably not worth the hassle to refactor, since as far as I can see we're not exposing that terminology to the end user anywhere?

cmcfadden commented on resources/js/pages/CoursePlanningPage/components/PersonTable/PersonTableCell.vue:
```diff
@@ -0,0 +1,192 @@
+});
+
+function handleSaveTentativeCourse({ term, course, person, role }) {
+  console.log("handleSaveTentativeCourse");
```
there are a few console.logs in this PR - leaving those in?

jxjj replied:
it's an aesthetic choice.  😄 

Actually, there is one console.log that I'd like to leave in (not this one). I'm logging some performance metrics on report page: the time from api call completion to table mounting, which should give a good estimate of the time it takes to compute. Things like changing from v-show to v-if, or precomputing a lookup table for a getter seem to have a significant impact on performance.

cmcfadden reviewed (approved):
looks good, excited to get this in front of some people

jxjj commented:
updates
- removed extraneous console.log
- fixed an issue where a mismatch between string emplid (from ldap) and int emplid (from db) could trigger a 500 response on first load, but normal responses on subsequent loads. https://github.com/UMN-LATIS/bluesheet/pull/79/commits/767e43e1399a67d0d30e02797fb6aed17d2a7127
- added a cached ldap lookup method to fix slow api responses for some depts https://github.com/UMN-LATIS/bluesheet/pull/79/commits/eec011b0a4740b73c3352cd02098fb040c966e3c
- changed sail's default cache to redis (while troubleshooting caching). Also, I learned that Laravel will interpret a null from `Cache::remember`  as a cache miss, rather than returning the cached `null` response.
- fixed tests

Known issues:
- laggy drag-and-drop for larger departments
